File: src/api/requirement.ts
import { invoke, PageResult } from "@/api";

export interface RequirementItem {
  id: number;
  name: string;
  requirement_code?: string;
  environment?: string;
  link?: string;
  iteration_id: number;
  status?: string;
  branch?: string;
  project_ids: number[];
  project_names: string[];
  sql_count: number;
  commit_count: number;
  created_at: string;
  updated_at: string;
  state: number;
}

export interface RequirementAddPayload {
  iteration_id: number;
  name: string;
  requirement_code?: string;
  environment?: string;
  link?: string;
  project_ids?: number[];
  status?: string;
  branch?: string;
}

export interface RequirementUpdatePayload {
  id: number;
  name?: string;
  requirement_code?: string;
  environment?: string;
  link?: string;
  project_ids?: number[];
  status?: string;
  branch?: string;
}

export interface RequirementStatusUpdatePayload {
  id: number;
  status: string;
}

export interface RequirementLinkPayload {
  requirement_id: number;
  link_type: string;
  link_id: number;
}

export interface RequirementCommitItem {
  id: number;
  project_id: number;
  project_name: string;
  commit_id: string;
  message: string;
  author_name: string;
  author_email: string;
  committed_at: string;
  additions: number;
  deletions: number;
  branch: string;
}

export interface RequirementCommitListReq {
  requirement_id: number;
  page?: number;
  size?: number;
  start_date?: string;
  end_date?: string;
}

export const requirementApi = {
  list: (iterationId: number) =>
    invoke<RequirementItem[]>("list_requirements", { req: { iteration_id: iterationId } }),
  listPage: (params: { iteration_id: number; page?: number; size?: number; status?: string; keyword?: string }) =>
    invoke<PageResult<RequirementItem>>("list_requirements_page", { req: params }),
  listCommits: (params: RequirementCommitListReq) =>
    invoke<PageResult<RequirementCommitItem>>("list_requirement_commits", { req: params }),
  add: (payload: RequirementAddPayload) =>
    invoke<number>("add_requirement", { req: payload }),
  update: (payload: RequirementUpdatePayload) =>
    invoke<void>("update_requirement", { req: payload }),
  updateStatus: (payload: RequirementStatusUpdatePayload) =>
    invoke<void>("update_requirement_status", { req: payload }),
  remove: (id: number) =>
    invoke<void>("delete_requirement", { id }),
  link: (payload: RequirementLinkPayload) =>
    invoke<void>("link_requirement", { req: payload }),
  linked: (linkType: string, linkId: number) =>
    invoke<RequirementItem | null>("get_linked_requirement", { linkType, linkId }),
};

File: src-tauri/src/models/requirement.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Requirement {
    pub id: i32,
    pub iteration_id: i32,
    pub name: String,
    pub requirement_code: String,
    pub environment: String,
    pub link: String,
    pub status: String,
    pub branch: String,
    pub state: i32,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct RequirementProject {
    pub id: i32,
    pub requirement_id: i32,
    pub project_id: i32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequirementListReq {
    pub iteration_id: i32,
    #[serde(default)]
    pub status: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequirementPageReq {
    pub iteration_id: i32,
    #[serde(default)]
    pub page: Option<i64>,
    #[serde(default)]
    pub size: Option<i64>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub keyword: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequirementAddReq {
    pub iteration_id: i32,
    pub name: String,
    #[serde(default)]
    pub requirement_code: Option<String>,
    #[serde(default)]
    pub environment: Option<String>,
    #[serde(default)]
    pub link: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub branch: Option<String>,
    #[serde(default)]
    pub project_ids: Option<Vec<i32>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequirementUpdateReq {
    pub id: i32,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub requirement_code: Option<String>,
    #[serde(default)]
    pub environment: Option<String>,
    #[serde(default)]
    pub link: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub branch: Option<String>,
    #[serde(default)]
    pub project_ids: Option<Vec<i32>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequirementStatusUpdateReq {
    pub id: i32,
    pub status: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequirementDeleteReq {
    pub id: i32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequirementMigrateReq {
    pub requirement_ids: Vec<i32>,
    pub target_iteration_id: i32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequirementLinkReq {
    pub requirement_id: i32,
    pub link_type: String,
    pub link_id: i32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequirementDetailRsp {
    #[serde(flatten)]
    pub requirement: Requirement,
    pub project_ids: Vec<i32>,
    pub project_names: Vec<String>,
    pub sql_count: i64,
    pub commit_count: i64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequirementCommitListReq {
    pub requirement_id: i32,
    #[serde(default)]
    pub page: Option<i64>,
    #[serde(default)]
    pub size: Option<i64>,
    #[serde(default)]
    pub start_date: Option<String>,
    #[serde(default)]
    pub end_date: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct RequirementCommitRsp {
    pub id: i32,
    pub project_id: i32,
    pub project_name: String,
    pub commit_id: String,
    pub message: String,
    pub author_name: String,
    pub author_email: String,
    pub committed_at: String,
    pub additions: i32,
    pub deletions: i32,
    pub branch: String,
}

